### ISSUE TITLE
OP_DATA argument A gets treated as register

### ISSUE THREAD

**Issue by Surasia** [Jake-NotTheMuss/hksc#4]
When decompiling a file, the OP_DATA instruction seems to be put into the second decompilation pass, causing its "A" value to be checked (by `isregvalid(fs, reg)`, which is usually above the max stack size.

```
#6  0x0000555555566040 in getexpinreg2 (fs=0x7fffffffcf20, reg=19) at ldecomp.c:4448
#7  0x000055555556cfcf in initexp2 (fs=0x7fffffffcf20, exp=0x7fffffffccd0, reg=20, pc=3) at ldecomp.c:6313
#8  0x000055555556d274 in addexp2 (sa=0x7fffffffce70, fs=0x7fffffffcf20, pc=3, o=OP_DATA, a=20, b=0, c=3, bx=3) at ldecomp.c:6361
#9  0x0000555555572d51 in blnode2 (sa=0x7fffffffce70, fs=0x7fffffffcf20, node=0x5555555a3c68) at ldecomp.c:8069
#10 0x0000555555573844 in pass2 (f=0x55555559f738, fs=0x7fffffffcf20) at ldecomp.c:8241
#11 0x000055555557398b in DecompileFunction (D=0x7fffffffd170, f=0x55555559f738) at ldecomp.c:8260
#12 0x0000555555573d58 in f_decompiler (H=0x55555559f2b8, ud=0x7fffffffd170) at ldecomp.c:8341
#13 0x0000555555574471 in luaD_rawrunprotected (H=0x55555559f2b8, f=0x555555573cc3 <f_decompiler>, ud=0x7fffffffd170) at ldo.c:106
#14 0x00005555555744d3 in luaD_pcall (H=0x55555559f2b8, func=0x555555573cc3 <f_decompiler>, u=0x7fffffffd170) at ldo.c:118
#15 0x0000555555573ed5 in luaU_decompile (H=0x55555559f2b8, f=0x55555559f738, w=0x555555558d3d <writer_2file>, data=0x5555555a03f0) at ldecomp.c:8365
#16 0x000055555555a4ea in lua_decompile (H=0x55555559f2b8, w=0x555555558d3d <writer_2file>, data=0x5555555a03f0) at lapi.c:270
#17 0x000055555555922a in hksc_dump_decomp (H=0x55555559f2b8, filename=0x7fffffffe00a "/home/./Downloads/thing.bin") at hkscaux.c:250
#18 0x00005555555587ba in hksc_dump_f (H=0x55555559f2b8, ud=0x7fffffffe00a) at hksc.c:575
#19 0x0000555555584635 in fparser (H=0x55555559f2b8, ud=0x7fffffffda20) at hksclib.c:297
#20 0x0000555555574471 in luaD_rawrunprotected (H=0x55555559f2b8, f=0x55555558451e <fparser>, ud=0x7fffffffda20) at ldo.c:106
#21 0x00005555555744d3 in luaD_pcall (H=0x55555559f2b8, func=0x55555558451e <fparser>, u=0x7fffffffda20) at ldo.c:118
#22 0x0000555555584877 in hksI_parser_file (H=0x55555559f2b8, filename=0x7fffffffe00a "/home/./Downloads/thing.bin", dumpf=0x5555555586d0 <hksc_dump_f>, 
    ud=0x7fffffffe00a) at hksclib.c:331
#23 0x000055555555883b in dofiles (H=0x55555559f2b8, argc=1, argv=0x7fffffffdc40) at hksc.c:588
#24 0x0000555555558a1c in main (argc=1, argv=0x7fffffffdc40) at hksc.c:642
```

**Comment by Jake-NotTheMuss** [comment]
Are you using an older version, because I updated the opcode properties at some point which prevented OP_DATA from being seen as an A-mode instruction.
The commit where I updated it is efd39cb. I will also make a tag for the last commit that has a mostly-working decompiler before I started rewriting it, in case you're interested.

**Comment by Surasia** [comment]
Yep, I was on commit [`44ff8a7`](https://github.com/Jake-NotTheMuss/hksc/commit/cbcf6c4e2ae211c123d8ad35cee1c2352ba8e07e). Just pulled the latest changes and the same still occurs.

edit: checked out the tag as well, seems like it has the same issue. Pretty sure the OP_DATA opcode isn't checked before being sent to `addexp2` function.

**Comment by Jake-NotTheMuss** [comment]
Yeah, I noticed that its "testAMode(o) || o == OP_DATA", which I added after changing the a-mode property of OP_DATA because local variables still need to be updated if its OP_DATA. I added the fix, but know that the current decompiler is incomplete and won't work on programs that have any sort of control structure. I am in the process of rewriting it. I made a tag that has the last functional version of the decompiler, which may be more helpful to you.
